File: src/main/java/fi/finlit/edith/ui/services/hibernate/DocumentNoteDaoImpl.java
/*
 * Copyright (c) 2009 Mysema Ltd.
 * All rights reserved.
 *
 */
package fi.finlit.edith.ui.services.hibernate;


import static fi.finlit.edith.sql.domain.QDocumentNote.documentNote;

import java.util.List;

import org.apache.tapestry5.ioc.annotations.Inject;
import org.springframework.util.Assert;

import fi.finlit.edith.sql.domain.Document;
import fi.finlit.edith.sql.domain.DocumentNote;
import fi.finlit.edith.sql.domain.Note;
import fi.finlit.edith.sql.domain.User;
import fi.finlit.edith.ui.services.DocumentNoteDao;
import fi.finlit.edith.ui.services.ServiceException;
import fi.finlit.edith.ui.services.UserDao;

public class DocumentNoteDaoImpl extends AbstractDao<DocumentNote> implements
        DocumentNoteDao {

    private final UserDao userDao;

    public DocumentNoteDaoImpl(@Inject UserDao userDao) {
        this.userDao = userDao;
    }

    @Override
    public List<DocumentNote> getOfDocument(Document document) {
        Assert.notNull(document);
        return query()
                .from(documentNote)
                .where(documentNote.document.eq(document),
                        // FIXME: Commented out, is good?
//                       documentNote.revision.loe(),
                       documentNote.deleted.eq(false))
                .orderBy(documentNote.position.asc()).list(documentNote);
    }

    @Override
    public List<DocumentNote> getPublishableNotesOfDocument(Document document) {
        return query()
            .from(documentNote)
            .where(documentNote.document.eq(document),
                    // FIXME: Commented out, is good?
//                   documentNote.reevision.loe(docRevision.getRevision()),
                   documentNote.deleted.isFalse(),
                   documentNote.publishable.isTrue())
            .orderBy(documentNote.position.asc()).list(documentNote);
    }

    @Override
    //XXX This is not really used anywhere?
    public void remove(DocumentNote docNote) {
        Assert.notNull(docNote, "note was null");
        // XXX What was the point in having .createCopy?
        docNote.setDeleted(true);
        docNote.getNote().decDocumentNoteCount();
        //getSession().save(docNote);
    }

    @Override
    public DocumentNote save(DocumentNote docNote) {
        if (docNote.getNote() == null) {
            throw new ServiceException("Note was null for " + docNote);
        }
        User createdBy = userDao.getCurrentUser();
        long currentTime = System.currentTimeMillis();
        docNote.setCreatedOn(currentTime);
        docNote.getNote().setEditedOn(currentTime);
        docNote.getNote().setLastEditedBy(createdBy);
        docNote.getNote().getAllEditors().add(createdBy);
        //getSession().save(docNote.getNote());
        //getSession().save(docNote);
        // FIXME: Hibernatify!
        //if (docNote.getNote().getComments() != null) {
        //    for (NoteComment comment : docNote.getNote().getComments()) {
        //        getSession().save(comment);
        //    }
        //}
        return docNote;
    }

    @Override
    public List<DocumentNote> getOfNote(Long noteId) {
        Assert.notNull(noteId);
        return query()
            .from(documentNote)
            .where(documentNote.note.id.eq(noteId),
                   documentNote.deleted.isFalse()).list(documentNote);
    }


    @Override
    public int getDocumentNoteCount(Note note) {
        return note.getDocumentNoteCount();
    }

    @Override
    public long getNoteCountForDocument(Long id) {
        return query()
            .from(documentNote)
            .where(documentNote.document.id.eq(id),
                   documentNote.deleted.isFalse())
            .count();
    }

    @Override
    public List<DocumentNote> getOfTerm(Long termId) {
        Assert.notNull(termId);
        return query()
            .from(documentNote)
            .where(documentNote.note.term.id.eq(termId),
                   documentNote.deleted.isFalse()).list(documentNote);
    }

    @Override
    public List<DocumentNote> getOfPerson(Long personId) {
        Assert.notNull(personId);
        return query()
            .from(documentNote)
            .where(documentNote.note.person.id.eq(personId),
                   documentNote.deleted.isFalse()).list(documentNote);
    }

    @Override
    public List<DocumentNote> getOfPlace(Long placeId) {
        Assert.notNull(placeId);
        return query()
            .from(documentNote)
            .where(documentNote.note.place.id.eq(placeId),
                   documentNote.deleted.isFalse()).list(documentNote);
    }

    @Override
    public DocumentNote getById(Long id) {
        return (DocumentNote) getSession().get(DocumentNote.class, id);
    }

}

File: src/main/scala/fi/finlit/edith/ui/services/hibernate/DocumentNoteDaoImpl.scala
package fi.finlit.edith.ui.services.hibernate

import fi.finlit.edith.sql.domain.QDocumentNote.documentNote
import java.util.List
import org.apache.tapestry5.ioc.annotations.Inject
import org.springframework.util.Assert
import fi.finlit.edith.sql.domain.Document
import fi.finlit.edith.sql.domain.DocumentNote
import fi.finlit.edith.sql.domain.Note
import fi.finlit.edith.sql.domain.User
import fi.finlit.edith.ui.services.DocumentNoteDao
import fi.finlit.edith.ui.services.ServiceException
import fi.finlit.edith.ui.services.UserDao
//remove if not needed
import scala.collection.JavaConversions._

class DocumentNoteDaoImpl(@Inject userDao: UserDao) extends AbstractDao[DocumentNote] with DocumentNoteDao {

  override def getOfDocument(document: Document): List[DocumentNote] = {
    Assert.notNull(document)
    query.from(documentNote).where(documentNote.document eq document, documentNote.deleted eq false).orderBy(documentNote.createdOn.asc()).list(documentNote)
  }

  override def getPublishableNotesOfDocument(document: Document): List[DocumentNote] = {
    query.from(documentNote).where(documentNote.document eq document, documentNote.deleted.isFalse, documentNote.publishable.isTrue).orderBy(documentNote.position.asc()).list(documentNote)
  }

  override def remove(docNote: DocumentNote) {
    Assert.notNull(docNote, "note was null")
    docNote.setDeleted(true)
    docNote.getNote.decDocumentNoteCount()
  }

  override def save(docNote: DocumentNote): DocumentNote = {
    if (docNote.getNote == null) {
      throw new ServiceException("Note was null for " + docNote)
    }
    var createdBy = userDao.getCurrentUser
    var currentTime = System.currentTimeMillis()
    docNote.setCreatedOn(currentTime)
    docNote.getNote.setEditedOn(currentTime)
    docNote.getNote.setLastEditedBy(createdBy)
    docNote.getNote.getAllEditors.add(createdBy)
    docNote
  }

  override def getOfNote(noteId: java.lang.Long): List[DocumentNote] = {
    Assert.notNull(noteId)
    query.from(documentNote).where(documentNote.note.id eq noteId, documentNote.deleted.isFalse).list(documentNote)
  }

  override def getDocumentNoteCount(note: Note): Int = note.getDocumentNoteCount

  override def getNoteCountForDocument(id: java.lang.Long): Long = {
    query.from(documentNote).where(documentNote.document.id eq id, documentNote.deleted.isFalse).count()
  }

  override def getOfTerm(termId: java.lang.Long): List[DocumentNote] = {
    Assert.notNull(termId)
    query.from(documentNote).where(documentNote.note.term.id eq termId, documentNote.deleted.isFalse).list(documentNote)
  }

  override def getOfPerson(personId: java.lang.Long): List[DocumentNote] = {
    Assert.notNull(personId)
    query.from(documentNote).where(documentNote.note.person.id eq personId, documentNote.deleted.isFalse).list(documentNote)
  }

  override def getOfPlace(placeId: java.lang.Long): List[DocumentNote] = {
    Assert.notNull(placeId)
    query.from(documentNote).where(documentNote.note.place.id eq placeId, documentNote.deleted.isFalse).list(documentNote)
  }

  override def getById(id: java.lang.Long): DocumentNote = {
    getSession.get(classOf[DocumentNote], id).asInstanceOf[DocumentNote]
  }
}
